File: implementations/go/internal/deltagen/delta.go
package deltagen

import (
    "sort"
)

// Operation types and JSON tags
type Operation struct {
    Op          string  `json:"op"`
    Path        string  `json:"path,omitempty"`
    From        string  `json:"from,omitempty"`
    To          string  `json:"to,omitempty"`
    ContentHash string  `json:"content_hash,omitempty"`
    Mode        *uint32 `json:"mode,omitempty"`
    Mtime       *int64  `json:"mtime,omitempty"`
    OldHash     string  `json:"old_hash,omitempty"`
    NewHash     string  `json:"new_hash,omitempty"`
    Content     string  `json:"content,omitempty"`
    Target      string  `json:"target,omitempty"`
}

// ComputeDelta produces a deterministic patch from src -> dst
func ComputeDelta(src Snapshot, dst Snapshot) ([]Operation, error) {
    srcMap := map[string]Entry{}
    dstMap := map[string]Entry{}
    for _, e := range src.Entries {
        srcMap[e.Path] = e
    }
    for _, e := range dst.Entries {
        dstMap[e.Path] = e
    }

    var ops []Operation
    var metadataOps []Operation

    // collect src-only and dst-only
    srcOnly := []string{}
    dstOnly := []string{}
    for p := range srcMap {
        if _, ok := dstMap[p]; !ok {
            srcOnly = append(srcOnly, p)
        }
    }
    for p := range dstMap {
        if _, ok := srcMap[p]; !ok {
            dstOnly = append(dstOnly, p)
        }
    }

    // same-path handling
    for p, se := range srcMap {
        if de, ok := dstMap[p]; ok {
            if se.Type != de.Type {
                // type changed: delete old, create new
                switch se.Type {
                case File:
                    ops = append(ops, Operation{Op: "delete_file", Path: p})
                case Dir:
                    ops = append(ops, Operation{Op: "delete_dir", Path: p})
                case Symlink:
                    ops = append(ops, Operation{Op: "delete_file", Path: p})
                }
                switch de.Type {
                case File:
                    ops = append(ops, Operation{Op: "create_file", Path: p, ContentHash: de.Hash})
                case Dir:
                    ops = append(ops, Operation{Op: "create_dir", Path: p})
                case Symlink:
                    ops = append(ops, Operation{Op: "symlink", Path: p, Target: de.LinkTarget})
                }
                continue
            }
            // same type
            if se.Type == File {
                if se.Hash != de.Hash {
                    metadataOps = append(metadataOps, Operation{Op: "modify_file", Path: p, OldHash: se.Hash, NewHash: de.Hash})
                } else {
                    if se.Mode != de.Mode {
                        m := uint32(de.Mode)
                        metadataOps = append(metadataOps, Operation{Op: "chmod", Path: p, Mode: &m})
                    }
                    if se.Mtime != de.Mtime {
                        mt := de.Mtime
                        metadataOps = append(metadataOps, Operation{Op: "utimes", Path: p, Mtime: &mt})
                    }
                }
            } else if se.Type == Dir {
                if se.Mode != de.Mode {
                    m := uint32(de.Mode)
                    metadataOps = append(metadataOps, Operation{Op: "chmod", Path: p, Mode: &m})
                }
            } else if se.Type == Symlink {
                if se.LinkTarget != de.LinkTarget {
                    metadataOps = append(metadataOps, Operation{Op: "symlink", Path: p, Target: de.LinkTarget})
                }
            }
        }
    }

    // build hash maps for move detection
    srcHash := map[string][]string{}
    dstHash := map[string][]string{}
    srcSize := map[string]int64{}
    for _, p := range srcOnly {
        if e, ok := srcMap[p]; ok && e.Type == File && e.Hash != "" {
            srcHash[e.Hash] = append(srcHash[e.Hash], p)
            srcSize[p] = e.Size
        }
    }
    for _, p := range dstOnly {
        if e, ok := dstMap[p]; ok && e.Type == File && e.Hash != "" {
            dstHash[e.Hash] = append(dstHash[e.Hash], p)
        }
    }

    // candidate pairs
    type pair struct {
        from string
        to   string
        size int64
    }
    var pairs []pair
    for h, dsts := range dstHash {
        if srcs, ok := srcHash[h]; ok {
            for _, d := range dsts {
                for _, s := range srcs {
                    pairs = append(pairs, pair{from: s, to: d, size: srcSize[s]})
                }
            }
        }
    }
    // sort by size desc, then from lexicographic
    sort.Slice(pairs, func(i, j int) bool {
        if pairs[i].size != pairs[j].size {
            return pairs[i].size > pairs[j].size
        }
        if pairs[i].from != pairs[j].from {
            return pairs[i].from < pairs[j].from
        }
        return pairs[i].to < pairs[j].to
    })

    usedSrc := map[string]bool{}
    usedDst := map[string]bool{}
    var selected []pair
    for _, pr := range pairs {
        if usedSrc[pr.from] || usedDst[pr.to] {
            continue
        }
        usedSrc[pr.from] = true
        usedDst[pr.to] = true
        selected = append(selected, pr)
    }

    // emit moves sorted by from
    sort.Slice(selected, func(i, j int) bool { return selected[i].from < selected[j].from })
    for _, s := range selected {
        ops = append(ops, Operation{Op: "move", From: s.from, To: s.to})
    }

    // remove moved from srcOnly/dstOnly
    filter := func(list []string, used map[string]bool) []string {
        out := make([]string, 0, len(list))
        for _, p := range list {
            if !used[p] {
                out = append(out, p)
            }
        }
        return out
    }
    srcOnly = filter(srcOnly, usedSrc)
    dstOnly = filter(dstOnly, usedDst)

    // deletes: deepest-first
    sort.Slice(srcOnly, func(i, j int) bool {
        di := depth(srcOnly[i])
        dj := depth(srcOnly[j])
        if di != dj {
            return di > dj
        }
        return srcOnly[i] < srcOnly[j]
    })
    for _, p := range srcOnly {
        if e, ok := srcMap[p]; ok {
            switch e.Type {
            case File:
                ops = append(ops, Operation{Op: "delete_file", Path: p})
            case Dir:
                ops = append(ops, Operation{Op: "delete_dir", Path: p})
            case Symlink:
                ops = append(ops, Operation{Op: "delete_file", Path: p})
            }
        }
    }

    // creates: parents-first
    sort.Slice(dstOnly, func(i, j int) bool {
        di := depth(dstOnly[i])
        dj := depth(dstOnly[j])
        if di != dj {
            return di < dj
        }
        return dstOnly[i] < dstOnly[j]
    })
    for _, p := range dstOnly {
        if e, ok := dstMap[p]; ok {
            switch e.Type {
            case Dir:
                ops = append(ops, Operation{Op: "create_dir", Path: p})
            case File:
                var m *uint32
                if e.Mode != 0 {
                    mm := uint32(e.Mode)
                    m = &mm
                }
                var mt *int64
                if e.Mtime != 0 {
                    mt = &e.Mtime
                }
                ops = append(ops, Operation{Op: "create_file", Path: p, ContentHash: e.Hash, Mode: m, Mtime: mt})
            case Symlink:
                ops = append(ops, Operation{Op: "symlink", Path: p, Target: e.LinkTarget})
            }
        }
    }

    // append metadata ops deterministically (sort by op then path)
    sort.Slice(metadataOps, func(i, j int) bool {
        if metadataOps[i].Op != metadataOps[j].Op {
            return metadataOps[i].Op < metadataOps[j].Op
        }
        return metadataOps[i].Path < metadataOps[j].Path
    })
    ops = append(ops, metadataOps...)
    return ops, nil
}

func depth(p string) int {
    if p == "" {
        return 0
    }
    cnt := 0
    for _, r := range p {
        if r == '/' || r == '\\' {
            cnt++
        }
    }
    return cnt
}

// InvertPatch returns a best-effort inverse patch
func InvertPatch(patch []Operation) ([]Operation, error) {
    var inv []Operation
    for i := len(patch) - 1; i >= 0; i-- {
        op := patch[i]
        switch op.Op {
        case "create_file":
            inv = append(inv, Operation{Op: "delete_file", Path: op.Path})
        case "delete_file":
            inv = append(inv, Operation{Op: "create_file", Path: op.Path})
        case "move":
            inv = append(inv, Operation{Op: "move", From: op.To, To: op.From})
        case "modify_file":
            inv = append(inv, Operation{Op: "modify_file", Path: op.Path, OldHash: op.NewHash, NewHash: op.OldHash})
        case "create_dir":
            inv = append(inv, Operation{Op: "delete_dir", Path: op.Path})
        case "delete_dir":
            inv = append(inv, Operation{Op: "create_dir", Path: op.Path})
        case "chmod":
            inv = append(inv, Operation{Op: "chmod", Path: op.Path, Mode: op.Mode})
        case "utimes":
            inv = append(inv, Operation{Op: "utimes", Path: op.Path, Mtime: op.Mtime})
        case "symlink":
            inv = append(inv, Operation{Op: "delete_file", Path: op.Path})
        }
    }
    return inv, nil
}

File: implementations/rust/src/delta.rs
use crate::snapshot::{Entry, EntryType, Snapshot};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, HashMap, HashSet};

/// Operation kinds
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "op")]
pub enum Operation {
    #[serde(rename = "create_file")]
    CreateFile {
        path: String,
        content_hash: Option<String>,
        mode: Option<u32>,
        mtime: Option<u64>,
        content: Option<String>,
    },
    #[serde(rename = "delete_file")]
    DeleteFile { path: String },
    #[serde(rename = "modify_file")]
    ModifyFile {
        path: String,
        old_hash: Option<String>,
        new_hash: Option<String>,
        content: Option<String>,
    },
    #[serde(rename = "create_dir")]
    CreateDir { path: String },
    #[serde(rename = "delete_dir")]
    DeleteDir { path: String },
    #[serde(rename = "move")]
    Move { from: String, to: String },
    #[serde(rename = "chmod")]
    Chmod { path: String, mode: u32 },
    #[serde(rename = "utimes")]
    Utimes { path: String, mtime: u64, atime: Option<u64> },
    #[serde(rename = "symlink")]
    Symlink { path: String, target: String },
}

/// Compute a deterministic patch from src -> dst
pub fn compute_delta(src: &Snapshot, dst: &Snapshot) -> anyhow::Result<Vec<Operation>> {
    // Build path maps
    let src_map = src.path_map();
    let dst_map = dst.path_map();

    // Collect sets
    let mut ops: Vec<Operation> = Vec::new();

    // 1) Handle identical paths and modifications
    let mut src_only_paths: Vec<String> = Vec::new();
    let mut dst_only_paths: Vec<String> = Vec::new();

    for p in src_map.keys() {
        if !dst_map.contains_key(p) {
            src_only_paths.push(p.clone());
        }
    }
    for p in dst_map.keys() {
        if !src_map.contains_key(p) {
            dst_only_paths.push(p.clone());
        }
    }

    // For paths present in both, check for modifications or metadata changes
    let mut metadata_ops: Vec<Operation> = Vec::new();
    for (p, s_entry) in &src_map {
        if let Some(d_entry) = dst_map.get(p) {
            if s_entry.entry_type != d_entry.entry_type {
                // Type changed: delete old, create new
                match s_entry.entry_type {
                    EntryType::File => ops.push(Operation::DeleteFile { path: p.clone() }),
                    EntryType::Dir => ops.push(Operation::DeleteDir { path: p.clone() }),
                    EntryType::Symlink => ops.push(Operation::DeleteFile { path: p.clone() }),
                }
                match d_entry.entry_type {
                    EntryType::File => ops.push(Operation::CreateFile {
                        path: p.clone(),
                        content_hash: d_entry.hash.clone(),
                        mode: d_entry.mode,
                        mtime: d_entry.mtime,
                        content: None,
                    }),
                    EntryType::Dir => ops.push(Operation::CreateDir { path: p.clone() }),
                    EntryType::Symlink => ops.push(Operation::Symlink {
                        path: p.clone(),
                        target: d_entry.link_target.clone().unwrap_or_default(),
                    }),
                }
                continue;
            }

            match (&s_entry.entry_type, &d_entry.entry_type) {
                (EntryType::File, EntryType::File) => {
                    if s_entry.hash != d_entry.hash {
                        metadata_ops.push(Operation::ModifyFile {
                            path: p.clone(),
                            old_hash: s_entry.hash.clone(),
                            new_hash: d_entry.hash.clone(),
                            content: None,
                        });
                    } else {
                        // same content, maybe metadata changes
                        if s_entry.mode != d_entry.mode {
                            if let Some(m) = d_entry.mode {
                                metadata_ops.push(Operation::Chmod { path: p.clone(), mode: m });
                            }
                        }
                        if s_entry.mtime != d_entry.mtime {
                            if let Some(mtime) = d_entry.mtime {
                                metadata_ops.push(Operation::Utimes { path: p.clone(), mtime, atime: None });
                            }
                        }
                    }
                }
                (EntryType::Dir, EntryType::Dir) => {
                    // directories: maybe metadata changes (mode/mtime)
                    if s_entry.mode != d_entry.mode {
                        if let Some(m) = d_entry.mode {
                            metadata_ops.push(Operation::Chmod { path: p.clone(), mode: m });
                        }
                    }
                }
                (EntryType::Symlink, EntryType::Symlink) => {
                    if s_entry.link_target != d_entry.link_target {
                        metadata_ops.push(Operation::Symlink {
                            path: p.clone(),
                            target: d_entry.link_target.clone().unwrap_or_default(),
                        });
                    }
                }
                _ => {}
            }
        }
    }

    // 2) Detect moves for files by hash
    // Build hash -> paths for src-only and dst-only files
    let mut src_hash_map: HashMap<String, Vec<String>> = HashMap::new();
    let mut dst_hash_map: HashMap<String, Vec<String>> = HashMap::new();
    let mut src_file_sizes: HashMap<String, u64> = HashMap::new();

    for p in &src_only_paths {
        if let Some(e) = src_map.get(p) {
            if let EntryType::File = e.entry_type {
                if let Some(h) = &e.hash {
                    src_hash_map.entry(h.clone()).or_default().push(p.clone());
                    if let Some(sz) = e.size {
                        src_file_sizes.insert(p.clone(), sz);
                    }
                }
            }
        }
    }
    for p in &dst_only_paths {
        if let Some(e) = dst_map.get(p) {
            if let EntryType::File = e.entry_type {
                if let Some(h) = &e.hash {
                    dst_hash_map.entry(h.clone()).or_default().push(p.clone());
                }
            }
        }
    }

    // Candidate moves: for each hash present in both maps, pair up paths.
    // We'll greedily match by preferring larger source file sizes first (minimize bytes moved).
    let mut move_pairs: Vec<(String, String, u64)> = Vec::new(); // (from, to, size)
    for (hash, dst_paths) in &dst_hash_map {
        if let Some(src_paths) = src_hash_map.get(hash) {
            // produce all candidate pairs
            for dst_p in dst_paths {
                for src_p in src_paths {
                    let size = *src_file_sizes.get(src_p).unwrap_or(&0u64);
                    move_pairs.push((src_p.clone(), dst_p.clone(), size));
                }
            }
        }
    }

    // Sort candidate pairs: prefer larger size, then lexicographically smaller source path
    move_pairs.sort_by(|a, b| {
        b.2.cmp(&a.2)
            .then_with(|| a.0.cmp(&b.0))
            .then_with(|| a.1.cmp(&b.1))
    });

    // Greedy selection ensuring non-overlapping paths
    let mut used_src: HashSet<String> = HashSet::new();
    let mut used_dst: HashSet<String> = HashSet::new();
    let mut selected_moves: Vec<(String, String)> = Vec::new();

    for (src_p, dst_p, _) in move_pairs {
        if used_src.contains(&src_p) || used_dst.contains(&dst_p) {
            continue;
        }
        used_src.insert(src_p.clone());
        used_dst.insert(dst_p.clone());
        selected_moves.push((src_p.clone(), dst_p.clone()));
    }

    // Emit moves (deterministic order: sort by from path)
    selected_moves.sort_by(|a, b| a.0.cmp(&b.0));
    for (from, to) in &selected_moves {
        ops.push(Operation::Move { from: from.clone(), to: to.clone() });
    }

    // Remove moved paths from src_only_paths and dst_only_paths
    let moved_src: HashSet<String> = selected_moves.iter().map(|(s, _)| s.clone()).collect();
    let moved_dst: HashSet<String> = selected_moves.iter().map(|(_, d)| d.clone()).collect();

    src_only_paths.retain(|p| !moved_src.contains(p));
    dst_only_paths.retain(|p| !moved_dst.contains(p));

    // 3) Remaining deletes (delete children before parents)
    // Sort src_only_paths by depth descending, then lexicographically
    src_only_paths.sort_by(|a, b| {
        let da = a.matches('/').count();
        let db = b.matches('/').count();
        db.cmp(&da).then_with(|| a.cmp(b))
    });

    for p in &src_only_paths {
        if let Some(e) = src_map.get(p) {
            match e.entry_type {
                EntryType::File => ops.push(Operation::DeleteFile { path: p.clone() }),
                EntryType::Dir => ops.push(Operation::DeleteDir { path: p.clone() }),
                EntryType::Symlink => ops.push(Operation::DeleteFile { path: p.clone() }),
            }
        }
    }

    // 4) Remaining creates (parents before children)
    // Sort dst_only_paths by depth ascending
    dst_only_paths.sort_by(|a, b| {
        let da = a.matches('/').count();
        let db = b.matches('/').count();
        da.cmp(&db).then_with(|| a.cmp(b))
    });

    for p in &dst_only_paths {
        if let Some(e) = dst_map.get(p) {
            match e.entry_type {
                EntryType::Dir => ops.push(Operation::CreateDir { path: p.clone() }),
                EntryType::File => ops.push(Operation::CreateFile {
                    path: p.clone(),
                    content_hash: e.hash.clone(),
                    mode: e.mode,
                    mtime: e.mtime,
                    content: None,
                }),
                EntryType::Symlink => ops.push(Operation::Symlink {
                    path: p.clone(),
                    target: e.link_target.clone().unwrap_or_default(),
                }),
            }
        }
    }

    // 5) Finally, append metadata ops deterministically sorted
    metadata_ops.sort_by(|a, b| {
        let ka = op_sort_key(a);
        let kb = op_sort_key(b);
        ka.cmp(&kb)
    });
    ops.extend(metadata_ops);

    Ok(ops)
}

/// Deterministic key for sorting metadata ops
fn op_sort_key(op: &Operation) -> (u8, String) {
    match op {
        Operation::Chmod { path, .. } => (5, path.clone()),
        Operation::Utimes { path, .. } => (6, path.clone()),
        Operation::ModifyFile { path, .. } => (4, path.clone()),
        Operation::Symlink { path, .. } => (3, path.clone()),
        _ => (9, "".to_string()),
    }
}

/// Invert a patch (produce operations that undo the given patch)
pub fn invert_patch(patch: &[Operation]) -> anyhow::Result<Vec<Operation>> {
    // This inversion is conservative: it inverts structural ops in reverse order.
    // For create_file we emit delete_file; for delete_file we emit create_file with no content (best-effort).
    let mut inv: Vec<Operation> = Vec::new();
    for op in patch.iter().rev() {
        match op {
            Operation::CreateFile { path, content_hash, mode, mtime, content } => {
                inv.push(Operation::DeleteFile { path: path.clone() });
                // Note: we cannot reconstruct original content here.
            }
            Operation::DeleteFile { path } => {
                inv.push(Operation::CreateFile {
                    path: path.clone(),
                    content_hash: None,
                    mode: None,
                    mtime: None,
                    content: None,
                });
            }
            Operation::Move { from, to } => {
                inv.push(Operation::Move { from: to.clone(), to: from.clone() });
            }
            Operation::ModifyFile { path, old_hash, new_hash, content: _ } => {
                inv.push(Operation::ModifyFile {
                    path: path.clone(),
                    old_hash: new_hash.clone(),
                    new_hash: old_hash.clone(),
                    content: None,
                });
            }
            Operation::CreateDir { path } => inv.push(Operation::DeleteDir { path: path.clone() }),
            Operation::DeleteDir { path } => inv.push(Operation::CreateDir { path: path.clone() }),
            Operation::Chmod { path, mode } => inv.push(Operation::Chmod { path: path.clone(), mode: *mode }),
            Operation::Utimes { path, mtime, atime } => inv.push(Operation::Utimes { path: path.clone(), mtime: *mtime, atime: *atime }),
            Operation::Symlink { path, target: _ } => inv.push(Operation::DeleteFile { path: path.clone() }),
        }
    }
    Ok(inv)
}
